File: leetcode/0007. Reverse Integer/Solution.swift
// https://leetcode.com/problems/reverse-integer/

class Solution {
    func reverse(_ x: Int) -> Int {
        var result = 0
        let sign = x < 0 ? -1 : 1
        var x = abs(x)
        while x > 0 {
            result = 10 * result + x % 10
            x /= 10
        }

        let r = sign * result
        return (r < Int32.min || r > Int32.max) ? 0 : r
    }
}

let s = Solution()
assert(s.reverse(0) == 0)
assert(s.reverse(123) == 321)
assert(s.reverse(-123) == -321)
assert(s.reverse(120) == 21)
assert(s.reverse(1534236469) == 0)

File: leetcode/0007. Reverse Integer/Solution.py
# https://leetcode.com/problems/reverse-integer/

class Solution:
    def reverse(self, x: int) -> int:
        if not x:
            return x
        
        result = 0
        sign = x // abs(x)
        x = abs(x)
        while x > 0:
            result = 10 * result + x % 10
            x //= 10

        signed_result = sign * result
        not_32_bit = signed_result < -2**31 or signed_result > 2**31-1
        return 0 if not_32_bit else signed_result

s = Solution()
assert s.reverse(0) == 0
assert s.reverse(123) == 321
assert s.reverse(-123) == -321
assert s.reverse(120) == 21
assert s.reverse(1534236469) == 0
assert s.reverse(1563847412) == 0